### PR TITLE
`iac scan prepush` command

### PR DESCRIPTION
**Pull request by Paul-GitGuardian**
Command meant to be used in a pre-push git hook to scan iac files for vulnerabilities.

Some parts to focus on:
- moved some functions from `ggshield/cmd/secret/scan/prepush.py` to `ggshield/core/git_hooks/prepush.py` in order for them to be used both in secrets and iac (and maybe sca in the future)

### CODE REVIEW

File: tests/functional/utils.py [hunk 1]
```diff
@@ -64,3 +65,20 @@ def recreate_censored_content(content: str, matched_string: str) -> str:
...
+
+
+# Use this in iac tests until iac hooks are added to the install command
+# Meaning they are added to install.py
+def create_local_hook(
+    path: Path,
+    hook_type: str,
+) -> None:
+    """Create hook directory (if needed) and pre-commit/pre-push file."""
+    hook_dir_path = path / ".git/hooks"
```
Comment (gg-teolohrer) [inline tests/functional/utils.py:77]:
```suggestion
    hook_dir_path = path / ".git" / "hooks"
```

File: tests/functional/utils.py [hunk 2]
```diff
@@ -64,3 +65,20 @@ def recreate_censored_content(content: str, matched_string: str) -> str:
     """Applies ggshield censoring to any occurrence of `matched_string` inside
     `content`. Returns the censored string."""
     return content.replace(matched_string, recreate_censored_string(matched_string))
+
+
+# Use this in iac tests until iac hooks are added to the install command
+# Meaning they are added to install.py
+def create_local_hook(
+    path: Path,
```
Comment (gg-teolohrer) [inline tests/functional/utils.py:73]: Directly specifying the hook dir path would allow to reuse this function for pre-receive also.
In the pre-receive context, the hook is executed in the `.git/` directory.

File: tests/functional/utils.py [hunk 3]
```diff
@@ -64,3 +65,20 @@ def recreate_censored_content(content: str, matched_string: str) -> str:
...
+    hook_type: str,
+) -> None:
+    """Create hook directory (if needed) and pre-commit/pre-push file."""
+    hook_dir_path = path / ".git/hooks"
+    hook_dir_path.mkdir(parents=True, exist_ok=True)
+    hook_path = hook_dir_path / hook_type
+
+    with hook_path.open("w") as f:
+        f.write("#!/bin/sh\n")
```
Comment (gg-teolohrer) [inline tests/functional/utils.py:82]:
```suggestion
        f.write("#!/usr/bin/env sh\n")
```
Not entirely sure this is the better way. It would avoid edge cases where there is no /bin/sh file

**Review by gg-teolohrer — CHANGES_REQUESTED** [review]

File: ggshield/core/git_hooks/prepush.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+from ggshield.core.git_shell import git, is_valid_git_commit_ref
+from ggshield.core.text_utils import display_warning
+from ggshield.core.utils import EMPTY_SHA
+
+
+logger = logging.getLogger(__name__)
+
+
+OUTDATED_HOOK_MESSAGE = """The installed pre-push hook did not pass its command-line arguments to ggshield. This can cause the hook to fail if the name of the remote you are pushing to is not "origin".
```
Comment (cyrvan-b) [inline ggshield/core/git_hooks/prepush.py:14]: NIT: 🏗 There is a `ggshield/core/constants.py` file, should we use a similar file for this kind of constants within iac commands?

Comment (Paul-GitGuardian) [reply ggshield/core/git_hooks/prepush.py:14]: IMO these are specific to prepush operations and would not benefit from being shared in a constants file. What do you think @Walz?

**Review by GG-HH — CHANGES_REQUESTED** [review]

File: ggshield/cmd/iac/scan/diff.py [hunk 5]
```diff
@@ -94,22 +106,32 @@ def iac_scan_diff(
...
+            and is_iac_file_path(file)
+        ]
+
+        if len(modified_iac_files) == 0:
+            return IaCSkipScanResult()
+
+    reference_tar = (
+        get_iac_tar(directory, ref, exclusion_regexes)
+        if ref is not None
+        else get_empty_tar()
```
[the remark this answers is not in the document]
Comment (Paul-GitGuardian) [reply ggshield/cmd/iac/scan/diff.py:133]: This case comes up on the first push to a repository.
I think it is more coherent to display all vulnerabilities as new in this case.
This should not take much more time than a scan all, as the ref tar is empty.

File: ggshield/core/git_hooks/prepush.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        ]
+    )
+    ancestors = output.splitlines()
+
+    if ancestors:
+        return ancestors[0]
+    return None
+
+
+def collect_commits_from_stdin(remote_name: str) -> Tuple[str, str]:
```
[the remark this answers is not in the document]
Comment (Paul-GitGuardian) [reply ggshield/core/git_hooks/prepush.py:59]: Sorry for that.
All these functions are the same as previously in cmd/secret/scan/prepush.py
`collect_commits_refs` contains code extracted (and untouched) from the secret scan prepush command

File: ggshield/cmd/iac/scan/diff.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (GG-HH) [reply ggshield/cmd/iac/scan/diff.py:133]: Yes but it would simplify the code in ggshield a lot and it is a very specific case that shouldn't happen very often (never).

[the remark this answers is not in the document]
Comment (Paul-GitGuardian) [reply ggshield/cmd/iac/scan/diff.py]: Ok, removed it

File: ggshield/core/git_hooks/prepush.py (hunk 4 above)
Comment (Walz) [reply ggshield/core/git_hooks/prepush.py:14]: I think If it's only used in this file, it can stay here.

**Review by Walz — APPROVED** [review]
LGTM

**Review by GG-HH — APPROVED** [review]

**Review by gg-teolohrer — APPROVED** [review]
